### PR TITLE
add etcd template resolver support

### PR DESCRIPTION
**Pull request by KaymeKaydex**
- now you can resolve some etcd values with oc updater
- update go version from legacy 1.11 -> 1.12(needs by etcd client)

### CODE REVIEW

**Review by aleksey-mashanov — CHANGES_REQUESTED** [review]

File: updater/main.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package main
 
 import (
 	"flag"
+	"github.com/onlineconf/onlineconf/updater/v3/updater/config"
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go:5]: yes, perhaps I will also add a linter

File: updater/main.go [hunk 2]
```diff
@@ -18,27 +19,23 @@ import (
...
 	Variables      map[string]string
+	ResolvePlugins config.ResolvePluginsConfig `yaml:"resolve_plugins"`
 }
 
 func main() {
 	flag.Parse()
 	config := readConfigFile(*configFile)
 	u := updater.NewUpdater(*config)
-
+	if config.ResolvePlugins.Enable {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go:36]: yes, initially there was an idea to make an implementation through a plugin as a binary file, the main disadvantage is the need for an ecosystem and a minimal sdk, which is not advisable to do yet. I'll just rename it to resolvers

File: updater/updater/template_resolvers.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		log.Error().Err(err).Msg("cant init " + etcd.Name() + "plugin")
+	}
+}
+
+var IncludedPlugins []IResolverPlugin
+
+func TryResolveByResolverPlugins(key string) (resolved string, ok bool) {
+	for _, p := range IncludedPlugins {
+		if strings.HasPrefix(key, p.Prefix()) {
+			val, err := p.Resolve(context.TODO(), strings.TrimPrefix(key, p.Prefix()))
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go:31]: Yes, I will inherit the root

File: updater/updater/resolver-plugins/etcd/etcd.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	if len(resp.Kvs) == 0 {
+		return "", nil
+	}
+
+	return string(resp.Kvs[0].Value), nil
+}
+
+func (r *Resolver) Info() string {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/resolver-plugins/etcd/etcd.go:49]: initially, when creating a model of the resolver module, I want to capture in the logs what it is

File: updater/updater/template_resolvers.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	if err == nil {
+		IncludedPlugins = append(IncludedPlugins, etcd)
+	} else {
+		log.Error().Err(err).Msg("cant init " + etcd.Name() + "plugin")
+	}
+}
+
+var IncludedPlugins []IResolverPlugin
+
+func TryResolveByResolverPlugins(key string) (resolved string, ok bool) {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go:28]: convenience, to fix that also no one of the resolvers could not resolve, And not just an empty resolved string there. it doesn't matter if we decide to delete

File: updater/updater/template_resolvers.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: done

File: updater/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go]: done

File: updater/updater/template_resolvers.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package updater
+
+import (
+	"context"
+	"github.com/onlineconf/onlineconf/updater/v3/updater/resolver-plugins/etcd"
+	"github.com/rs/zerolog/log"
+	"strings"
+)
+
+type IResolverPlugin interface {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: fixed

File: updater/updater/admin.go [hunk 7]
```diff
@@ -140,7 +140,18 @@ func prepareModules(data *ConfigData, vars map[string]string) (modules map[strin
 		case "application/x-template":
 			mParam.value = templateRe.ReplaceAllStringFunc(param.Value.String, func(match string) string {
 				name := match[2 : len(match)-1]
-				return vars[name]
+
+				val := vars[name]
+				if vars[name] == "" {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/admin.go]: fixed

File: updater/updater/resolver-plugins/etcd/etcd.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/resolver-plugins/etcd/etcd.go]: i removed; done

File: updater/etc/onlineconf.yaml [hunk 8]
```diff
@@ -1,8 +1,16 @@
...
+  host: 127.0.0.1
   port: 80
-  username: onlineconf.updater
-  password: xxx
+  username: gopher
+  password: gopher
 data_dir: /usr/local/etc/onlineconf/
 reload_local: 10
 update_interval: 60
+resolve_plugins:
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/etc/onlineconf.yaml:9]: I think the example should also be an example of a local startup, the values of xxx are a bit confusing; in general, it does not matter to me

File: updater/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go]: done

File: updater/updater/template_resolvers.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: if no one plugin can resolve such value

[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: i added log if no resolver could get the value

File: updater/updater/template_resolvers.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	Resolve(context.Context, string) (string, error)
+	Info() string
+	Prefix() string
+	Name() string
+}
+
+func InitPluginsUsage(pluginsCfgs map[string]map[string]string) {
+	etcd, err := etcd.New(pluginsCfgs[etcd.ResolverName])
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go:18]: there is a log if the resolve module could not be started

File: updater/updater/template_resolvers.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+		IncludedPlugins = append(IncludedPlugins, etcd)
+	} else {
+		log.Error().Err(err).Msg("cant init " + etcd.Name() + "plugin")
+	}
+}
+
+var IncludedPlugins []IResolverPlugin
+
+func TryResolveByResolverPlugins(key string) (resolved string, ok bool) {
+	for _, p := range IncludedPlugins {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: separator is :
in what way predictable? it seems that giving the resolver to recognize the prefix itself is less optimal than asking the resolver if it is his prefix

**Comment by aleksey-mashanov** [comment]
Modifications of a variable in `etcd` must be propagated to the configuration files without requiring touching of OnlineConf tree.

**Review by aleksey-mashanov — CHANGES_REQUESTED** [review]

File: updater/etc/onlineconf.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/etc/onlineconf.yaml]: this case is just an example. it disables all resolvers by default

File: updater/updater/template_resolvers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: I rname it from ModuleResolver to TemplateVariableResolver

File: updater/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go]: I found one place where the name of the plugin remained and fixed it, revised it now, it doesn't seem to be left

File: updater/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go]: done, fixed the sorting in my ide that was getting in the way

File: updater/main.go [hunk 11]
```diff
@@ -18,29 +20,26 @@ import (
...
 	flag.Parse()
 	config := readConfigFile(*configFile)
 	u := updater.NewUpdater(*config)
-
+	if config.ResolveModules.Enable {
+		updater.InitResolveModulesUsage(config.ResolveModules.Modules)
+	}
 	if *once {
-		if u.Update() != nil {
+		if u.Update(ctx) != nil {
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go:42]: how is the root context able to change the work of the api?

File: updater/updater/template_resolvers.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: you mean plugin:plugin_name:key?

File: updater/etc/onlineconf.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/etc/onlineconf.yaml]: what's your suggestion?

File: updater/updater/template_resolvers.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: what's your suggestion?

[the remark this answers is not in the document]
Comment (aleksey-mashanov) [reply updater/updater/template_resolvers.go]: Die if an error emitted on the first run, use previous value in the other case.

File: updater/updater/template_resolvers.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (aleksey-mashanov) [reply updater/updater/template_resolvers.go]: `${plugin_name}:${key}` to be compatible with vault injection syntax.

File: updater/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (aleksey-mashanov) [reply updater/main.go:42]: u.Update() – is a part of the public API and is used by onlineconf-csi-driver.

File: updater/main.go [hunk 12]
```diff
@@ -18,27 +19,23 @@ import (
 var configFile = flag.String("config", "/usr/local/etc/onlineconf.yaml", "config file")
 var once = flag.Bool("once", false, "fetch configuration once and exit")
 
-type AdminConfig struct {
```
[the remark this answers is not in the document]
Comment (aleksey-mashanov) [reply updater/main.go:21]: No. They was intentionally split apart. `updater` is a public package and its AdminConfig is used to provide public API. AdminConfig in main is used for yaml-config parsing and must be backward compatible with previous versions to not force users to update config on minor version upgrade.

File: updater/etc/onlineconf.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/etc/onlineconf.yaml]: I deleted everything new

File: updater/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go]: removed root ctx, now ctx is context.TODO with todo comment

File: updater/main.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go]: I returned it to the old place

File: updater/updater/template_resolvers.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: I rplcd separator usage

File: updater/updater/template_resolvers.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: check please current decision

File: updater/updater/template_resolvers.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+func (c *ResolveModulesValChecker) CleanStorage() {
+	c.storage.clean()
+}
+
+func TryResolveByResolverModule(ctx context.Context, key string) (resolved string, ok bool) {
+	for _, m := range IncludedResolveModules {
+		if strings.HasPrefix(key, m.Name()+separator) { // that means that val created for resolver usage
+			resolverKey := strings.TrimPrefix(key, m.Name()+separator)
+			val, err := m.Resolve(ctx, resolverKey)
+			ResolveChecker.Put(m, resolverKey, val)
```
Comment (Nikolo) [inline updater/updater/template_resolvers.go:44]: А точно ли надо Put-ить значение в случае если произошла ошибка?

File: updater/main.go [hunk 14]
```diff
@@ -29,16 +31,27 @@ type ConfigFile struct {
...
+	ResolveModules config.ResolveModulesConfig `yaml:"resolve_modules"`
 }
 
 func main() {
+	ctx := context.Background()
 	flag.Parse()
 	config := readConfigFile(*configFile)
 	u := updater.NewUpdater(*config)
+	if config.ResolveModules.Enable {
+		updater.ResolveChecker = updater.NewResolveModulesValChecker(ctx, 10*time.Second, u.UpdateForce)
```
Comment (Nikolo) [inline updater/main.go:46]: Magic num 10
Может в конфиг унести?

File: updater/updater/template_resolvers.go [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+type kv struct {
+	key   string
+	value string
+}
+
+type ResolveModulesValChecker struct {
+	ctx             context.Context
+	rebuildTreeFunc func() error
+	cancel          context.CancelFunc
+	tickReady       atomic.Bool
```
Comment (Nikolo) [inline updater/updater/template_resolvers.go:66]: А зачем это поле? Предполагалось его использование?

File: updater/updater/template_resolvers.go [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+const separator = ":"
+
+type TemplateVariableResolver interface {
+	Resolve(context.Context, string) (string, error)
+	Name() string
+}
+
+func InitResolveModulesUsage(resolversCfgs map[string]map[string]string) error {
+	etcd, err := etcd.New(resolversCfgs[etcd.ResolverName])
```
Comment (Nikolo) [inline updater/updater/template_resolvers.go:23]: На данном этапе предполагается, что т.к. резолвер один то если они включены то по любому надо попытаться подключить etcd?

Может сразу попробовать про конструкцию со switch case-ом или как то так? Боюсь, что если в resolversCfgs приедет, какое то отличное от etcd значение, и при этом не приедет etcd, то ошибка может выглядеть очень странно или непример её вообще не будет если etcd.New, переварит пустую конфигурацию.

**Review by Nikolo — CHANGES_REQUESTED** [review]

File: updater/main.go [hunk 17]
```diff
@@ -29,16 +31,27 @@ type ConfigFile struct {
...
+		go updater.ResolveChecker.StartCronCheck()
```
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go:47]: поправлю

File: updater/updater/template_resolvers.go (hunk 13 above)
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go:44]: да, чтобы если оно пустое на данный момент и потом станет не пустым то оно само починилось

File: updater/main.go (hunk 14 above)
Comment (KaymeKaydex) [reply updater/main.go:46]: поправлю, забыл убрать после дебага

File: updater/updater/template_resolvers.go (hunk 15 above)
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go:66]: верно, поправлю

File: updater/updater/template_resolvers.go (hunk 16 above)
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go:23]: верно, поправлю

File: updater/main.go (hunk 14 above)
Comment (KaymeKaydex) [reply updater/main.go]: исправил на config.UpdateInterval

File: updater/updater/template_resolvers.go (hunk 15 above)
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: поправил, убрал

File: updater/main.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (KaymeKaydex) [reply updater/main.go]: пофиксил

File: updater/updater/template_resolvers.go (hunk 13 above)
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go:43]: поговорили на эту тему, закроем тред

File: updater/updater/template_resolvers.go (hunk 16 above)
Comment (KaymeKaydex) [reply updater/updater/template_resolvers.go]: добавил заглушкой с проверкой конфига на нил и todo: чтобы это не забыть сделать

**Pull request closed without merging**